src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1104,8 +1105,17 @@ UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
         kh_del(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, it);
     }
 
-    if ((self->dci == UCT_DC_MLX5_EP_NO_DCI) ||
-        uct_dc_mlx5_iface_is_dci_rand(iface)) {
+    if (uct_dc_mlx5_iface_is_dci_rand(iface) ||
+        (self->dci == UCT_DC_MLX5_EP_NO_DCI)) {
+        return;
+    }
+
+    pool_index = uct_dc_mlx5_iface_dci_pool_index(iface, self->dci);
+    if (uct_dc_mlx5_iface_dci_free(iface, self)) {
+        ucs_assert(self->dci == UCT_DC_MLX5_EP_NO_DCI);
+        if (!(iface->flags & UCT_DC_MLX5_IFACE_FLAG_IN_DCI_PROGRESS)) {
+            uct_dc_mlx5_iface_progress_pending(iface, pool_index);
+        }
```

Comment:
AS discussed, we should not call progress pending from this context. Pending callbacks should be called only from uct_worker_progress() context.
Can we use the fact that uct_dc_mlx5_ep_pending_purge() is scheduling a progress callback, to eventually progress the arbiter? for example, we could mark dci->ep=(void*)-1 to mean that the dci is still held and should be released by uct_dc_mlx5_ep_dci_release_progress

Reply:
ok, done: mark dci->ep=(void*)-1 as release candidate